### ISSUE TITLE
Support for the Go 1.5 Vendor Experiment (GO15VENDOREXPERIMENT)

### ISSUE THREAD

**Issue by Vektah** [go-lang-plugin-org/go-lang-idea-plugin#1820]
# Do not add "+1" comments
__Kindly do not add "+1" comments to this issue.__ 
Read [comment #1820](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1820#issuecomment-166231301)

The rest of us are subscribing to this issue waiting for real updates, not email messages about someone "expressing interest" in this issue.

------
Original issue:

It seems there is no direct support for looking doing package resolution into the vendor directory.

I am currently working around it by creating a `src` symlink in the vendor directory that points to itself and adding the vendor directory as a project GOPATH.
![symlink](https://cloud.githubusercontent.com/assets/2247982/9403270/e3920320-4828-11e5-8966-cfc1ad59ae61.png)

It would be nice if the plugin automatically detected the presence of the vendor directory and added it automatically.

When support for the vendor experiment is added this should be added to the other tools / run configurations as well in order to enable transparent compiling / running of those tools.

The corresponding environment variable to enables the experiment, `GO15VENDOREXPERIMENT=1`

Related: https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo

**Comment by yograterol** [comment]
+1

**Comment by shwoodard** [comment]
+1

**Comment by vburenin** [comment]
If you read carefully the description of first page, you would find a
section containing such text:

Bumping or +1 or me as well comments

As annoying as things might be when they don't work, please don't comment
with "bump", "+1", "same for me" or other form of comments which is not
relevant for either better identifying the issue providing a solution to it.

If you want to mark that the said issue is meaningful for you, go ahead and
contact Github support to add a issue voting feature to the repositories by
writing them here: https://github.com/contact

Contributions are always welcomed and we'll do our best to help you out to
add yours to the project.

**Comment by piotrkowalczuk** [comment]
Subscription is enough to show your interest.

**Comment by dlsniper** [comment]
@zolotov I've added the following to the issue description:
"When support for the vendor experiment is added this should be added to the other tools / run configurations as well in order to enable transparent compiling / running of those tools."

The initial issue revolved mostly around the importing of packages but in addition to packages, all the other tools in the project, `go get`, `go run`, `go build` for example, will need to have the `GO15VENDOREXPERIMENT=1` env var present at execution time.

**Comment by apokusin** [comment]
+1

**Comment by ansel1** [comment]
In addition, if you do the `src` symlink as suggested above, but *also* have a copy of those packages in your $GOPATH, the plugin gives errors saying "Resolved to several targets".

**Comment by mihaicostin** [comment]
I've done it other way around - create a symlink in my src folder 'github.com' pointing to 'vendor/github.com' - and it works ok. the downside to this is that you'll need to create a symlink for each vendored top package name.

**Comment by jstrachan** [comment]
If its any help, here's a little bash script to create a new `src` folder and then create a symlink in that folder to each folder in the vendor folder:
```
#!/bin/bash
rm -rf src
mkdir src
FILES=vendor/*
for f in $FILES
do
  FULL=`pwd`/$f
  echo "Symlinking vendor source dir: $FULL"
  ln -s $FULL src
done
```

**Comment by dlsniper** [comment]
@jstrachan thanks for sharing but I wouldn't recommend using it. The script removes the src folder which could be dangerous for some users. And the "better" workaround would be to create a folder called `src` in `<project dir>/vendor` (`<project dir>/vendor/src` is the final path) and then symlink everything there. Once that is done, the vendor directory should be usable as a GOPATH for the plugin without issues.

**Comment by jstrachan** [comment]
@dlsniper thanks for the heads up - I just updated the script to use vendor/src ;)

**Comment by jstrachan** [comment]
I guess the script will barf if you run it twice  as it'll link to itself

**Comment by dlsniper** [comment]
Nice. I guess people can realize there's no need to run it twice ;)

**Comment by jstrachan** [comment]
@dlsniper if your vendor dependencies bring in a new github project you'd need to rerun it again. Just polished the script some more so it can be run many times & it recreates the `vendor/src` each time

**Comment by gerred** [comment]
:+1: on this. I am *really* liking this plugin (even over vim-go), and it's making me use WebStorm because of it. I'd love a standalone Go IDE, but this'll do for now, the one thing I'm really wishing is that package resolution/cmd-click for packages in `vendor`

**Comment by ykhrustalev** [comment]
+1

**Comment by dlsniper** [comment]
@ykhrustalev / others I understand that the issue is of some sort of importance to you, but if you want to express that, please use the subscribe to issue or write to Github support to add a voting feature: https://github.com/go-lang-plugin-org/go-lang-idea-plugin#bumping-or-1-comments 

The issue is that whenever you send a +1 message you send a mail to all those who subscribe to the issue and in this case, 11 other people. Again, not your fault, Github should have a better way to handle this.

Thank you!

**Comment by samy100** [comment]
+1

**Comment by dlsniper** [comment]
@zolotov I've realized that there's one more case for this work correctly.

Given the following situation:
- package path: ` github.com/dlsniper/u `
- project directory ` <package root> `
- dependencies in ` <package root>/vendor `
- nothing in the Global libraries
- under Project libraries have the ` <project root>/vendor ` entry

When one creates a project either with GO15VENDOREXPERIMENT or another package manager like Godeps, then even if the dependencies are resolved correctly the plugin doesn't understand imports like this: ` import "github.com/dlsniper/u/subpackage" ` because it doesn't have a way to determine that the current path is actually part of the ` github.com/dlsniper/u ` package.

If then the user adds the GOPATH which contains ` github.com/dlsniper/u ` as a package, but other packages that are also in ` <project root>/vendor ` then the plugin will mark those as invalid / duplicate.

As such, the issue in this case would be having the Project libraries completely taking over the Global libraries when duplicates exists and suppressing them.

Let me know if I should file this as a separate issue / clarify it further as I know you've said you'd like to rework the vendor issue as part of some other changes as well.

Thank you.

**Comment by savalin** [comment]
+1

**Comment by PapaStifflera** [comment]
+1

**Comment by morphinewan** [comment]
+1

**Comment by d-ulyanov** [comment]
+1

**Comment by ignatov** [comment]
Please don't use +1, we know it's a major feature and we're going to implement it soon. Thanks.

**Comment by dlsniper** [comment]
@ignatov I think this is more voted these days because Go 1.6 introduces support for vendoring on by default while 1.7 will remove the soon to be deprecated flag, see the release notes: https://tip.golang.org/doc/go1.6#go_command

Fortunately there are no other changes, for the language or otherwise, that we should be concerned with.

I'll keep adding input here in regards to how I would see this working but I would expect others to give their input as well.

**Comment by vvelikodny** [comment]
+1

**Comment by DallanQ** [comment]
In case anyone else runs into the same issue I did, adding a `src` symlink pointing to `.` under the `vendor` directory resulted in a bunch of inspections failing with "Unresolved reference" to functions defined in vendor modules. (Curiously, not every function reference was marked unresolved; just a few.) So instead I created a `_vendor` directory as a sibling of `vendor`, added a `src` symlink pointing to `../vendor` under `_vendor`, and added `_vendor` as a project library path in Settings. That solved the problem for me.

**Comment by gus** [comment]
The `_vendor` suggestion from @DallanQ solved our problems ... until we updated `glide` today. `glide` now WARNs and then exits if you have a `_vendor` directory ... so we renamed `_vendor` to `_vendoor`.

Hoping this issue gets resolved in the plugin soon, though :)

**Comment by bsideup** [comment]
:+1:

**Comment by gobwas** [comment]
+1

**Comment by brainm** [comment]
+1

**Comment by dlsniper** [comment]
@ignatov / @zolotv would you consider locking this issue so that we don't get +1 spam?

**Comment by shwoodard** [comment]
People!  Seriously!  They've asked for no +1's and it is being ignored and now they are considering locking the issue, and I can't say I blame them.  But, this impedes collaboration.  Sad.  Follow/subscribe to the issue if you're interested!  These are the instructions provided by the repo owners.  I expect that you prefer that other adhere to your requests on your oss.  Kindly, return the courtesy.

**Comment by dlsniper** [comment]
@shwoodard the issue is not ignored, see what Sergey wrote above: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1820#issuecomment-166231301. It's a bit complex, see my previous comments about the building, formatter and so on and it takes a bit of time to get all the other internals in place. And I do understand that +1 is just a form of showing that people need it but due to how @github works unfortunately, it makes is terrible when you get an e-mail with +1 (even after being in the README of the project and asked a couple of times here). Locking the issue would only be helpful not to get the +1 and allow discussion on the subject itself (and I feel terrible I had to suggest this).

**Comment by ganting** [comment]
+1

**Comment by sogko** [comment]
Is it possible for someone to update the original issue description and write a notice at the top it regarding commenting with +1s and its status? 

I can understand that some might have just read the original issue description, find themselves interested in it and didn't even read the following comments.

Hopefully this can be seen by the next would-be commentor, (who probably could be sitting in the [International Space Station](https://en.wikipedia.org/wiki/International_Space_Station))

# Do not add "+1" comments
__Kindly do not add "+1" comments to this issue.__ 
Read [comment #1820](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1820#issuecomment-166231301)

The rest of us are subscribing to this issue waiting for real updates, not email messages about someone "expressing interest" in this issue.

**Comment by dlsniper** [comment]
Done. Thanks for suggestion. I'll cleanup the issue soon as there are some other things that should be mentioned for it.

**Comment by asampal** [comment]
@gus what you wrote in [this comment](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1820#issuecomment-175233696) is working out OK for you for the time being? I'm considering using vendoring with glide, but the current situation wrt this plugin makes me hesitate.

**Comment by mihaicostin** [comment]
@asampal I'm using successfully glide with the approach I've mentioned above (symlink in the src folder to the root folder from vendor, e.g. github.com). There is the downside of having a symlink for each "domain", but it will do until this gets fixed. 
And yes.. the +1 thing is really really annoying. It gets my hopes high whenever I get a new message from this topic in the inbox :)

**Comment by chriswhelix** [comment]
These are the steps I use, only requires doing it once per project:

1. In the Project Explorer, right-click the ‘vendor’ directory and choose Mark As -> Excluded.

2. At the command line, in the project dir, run:
`
mkdir _idea_vendor
ln -s ../vendor _idea_vendor/src
`

3. In Preferences / Languages / Go / Go Libraries, add a Project Library of _idea_vendor/src

This works for me for package name completion stuff in the editor. I haven't tried any of the run/build tools with this configuration, I tend to do that outside the IDE.

**Comment by dlsniper** [comment]
@zolotov I'm not sure how much you've looked into this so far (I've seen the branch but I'm having terrible time issues). There's one more thing for the vendoring stuff. From 1.7 the GO15VENDOREXPERIMENT will disappear but the functionality will be present ENABLED by default. Right now, in 1.6 it's turned ON by default and it was OFF in 1.5 (prior it didn't existed).

Also, it's worth noting that there are two types of projects:
- applications (usually people might want vendoring turned on for the project)
- libraries (normally vendoring should be turned on but the vendor folder not committed to the repository).

I'm not sure how would be the best way to have this done, maybe ask the user on a per-project basis if they want vendoring on / off and allow them to change their minds later on?

Also, I'm thinking that later on we might want to allow the plugin to use vendoring tools such as `glide`, `godep`, `govendor` etc (these are all kinda doing the same thing, as opposed to `gb` which changes the project layout and some other things, there's an issue for it as well so maybe `x.x.x pool` as it's marked right now is a good target) so this might be worth considering, if possible, while solving this.

I hope this helps and thank you so much for the work on the plugin (to Sergey as well, I know I don't say it nearly enough).

**Comment by zolotov** [comment]
Hello everyone,

I've pushed an implementation of vendoring support. The last builds in nightly channel (0.11.1192, 0.10.1189, 0.9.1191) contain property imports resolution algorithm for vendored packages, some new inspections for imports and some package completion cleaning up. 

Please update the plugin, restore the initial project configuration (get rid of workarounds like symlinks and vendor-directory exclusions) and try it out.

Any feedback is welcome. Especially I'd appreciate if you report issues about redcode in imports, redundant or unreachable packages that are appeared in completion or auto-import popup.

NOTE: for Go 1.5 version you have to enable vendoring explicitely for module via `File | Settings | Languages & Frameworks | Go | Project Settings`

**Comment by F21** [comment]
Really awesome stuff! Not entirely sure if it's due to this change, but I noticed if I want to use the standard library and type something like `net.P` and press enter after selecting the correct method, it doesn't import the `net` package.

**Comment by dlsniper** [comment]
@F21 I can't replicate this. Can you please open up a new issue and specify your Go project settings (Settings | Languages & Frameworks | Go | Project Settings) and make sure that your project and module both use the correct Go SDK? Thank you.

**Comment by dcelasun** [comment]
Thanks for the great work guys! Tested with 0.10.1189 on IDEA 15.0.4, everything seems to work fine. No problems whatsoever with imports, auto-import popups, ctrl+click or redundant packages. I also can't reproduce @F21's problem.

**Comment by onceatime** [comment]
awesome! THX gentleman this is a huge step... can't say it loud enough

**Comment by DallanQ** [comment]
Agreed - thank you so much!

**Comment by skovtunenko** [comment]
@zolotov , for me work well! Now we can drop 'src' folder with symlinks from inside 'vendor' folder.

**Comment by jeromelaurens** [comment]
@zolotov , work for me.
Mac OSX
IDEA 2016.1 CE
GO 1.6
plugin version : 0.11.1205
I had to check the option in pref, language go, project settings, but that's fine
Thanks a lots for your great work !

**Comment by zolotov** [comment]
@jeromelaurens 

> I had to check the option in pref, language go, project settings, but that's fine

That's strange. What is the exactly version of your go installation? Please could you attach screenshot of `File | Project Settings | SDKs` with selected Go SDK and `File | Settings | Languages & Frameworks | Go | Project Settings` with selected `Default` for vendoring?

**Comment by jeromelaurens** [comment]
@zolotov 
go version : go1.6 darwin/amd64
![image](https://cloud.githubusercontent.com/assets/2656354/13953234/379f1d62-f03b-11e5-8c0f-6513c4545e53.png)
![image](https://cloud.githubusercontent.com/assets/2656354/13953258/52a14f68-f03b-11e5-8655-a867810dd991.png)

**Comment by zolotov** [comment]
@jeromelaurens IDEA thinks that you use go1.5. Maybe 1.5 version was cached by IDE and hasn't updated after upgrading go. To make sure, please attach content of `/usr/local/go/src/runtime/internal/sys/zversion.go`.

**Comment by jeromelaurens** [comment]
@zolotov the content is:
```
// auto generated by go tool dist

package sys

const DefaultGoroot = `/usr/local/go`
const TheVersion = `go1.6`
const Goexperiment = ``
const StackGuardMultiplier = 1
```

**Comment by cescoferraro** [comment]
Have you tried to set the vendor folder as an Intellij library root? it works for me

**Comment by dlsniper** [comment]
@cescoferraro what does that mean? With the latest plugin version (at least on nightly) you don't need to do any such thing. Please file an issue for that as it seems wrong to do what you suggested. Thank you.

**Comment by soerenmartius** [comment]
i had the same issue just today, i removed the the plugin from intellij as well as the .idea/ directory in my project and reinstalled the plugin, it worked afterwards.
